py-tests/pyspark_tests.py

```diff
@@ -0,0 +1,642 @@
+'''
+
+def test_spark_df_ts_write_read_one_entry(spark_context, docker_cli, riak_client, sql_context):
+
+    host, pb_port, hostAndPort = get_host_and_port(docker_cli)
+    temp_table, temp_table_name, created = create_table(riak_client)
+    assert created == True
+
+    test_row_in = ['field1_val', 'field2_val', datetime.datetime.fromtimestamp(1293840000), 0]
+
+    test_rdd = spark_context.parallelize([test_row_in])
+    test_df = test_rdd.toDF(['field1', 'field2', 'datetime', 'data'])
+
+    assert df_write_retry(test_df, temp_table_name) == True
+
+    query = """select * from %(table_name)s
+                    where datetime >= %(start_date)s
+                    AND datetime <=  %(end_date)s
+                    AND field1 = '%(field1)s'
+                    AND field2 = '%(field2)s'
+                """ % ({'table_name': temp_table_name, 'start_date': 1293840000000-1, 'end_date': 1293840000000+1, 'field1': 'field1_val', 'field2': 'field2_val'})
+
+    assert ts_write_verify(riak_client, temp_table_name, query, [[b'field1_val', b'field2_val', 1293840000000, 0]]) or \
+            ts_write_verify(riak_client, temp_table_name, query, [[b'field1_val', b'field2_val', datetime.datetime(2011, 1, 1, 0, 0), 0]]) == True
+
+    temp_filter = """datetime >= CAST(%(start_date)s AS TIMESTAMP)
+                    AND datetime <=  CAST(%(end_date)s AS TIMESTAMP)
+                    AND field1 = '%(field1)s'
+                    AND field2 = '%(field2)s'
+                """ % ({'start_date': 1293840000-1, 'end_date': 1293840000+1, 'field1': 'field1_val', 'field2': 'field2_val'})
+
+    read_df = sql_context.read.format("org.apache.spark.sql.riak").load(temp_table_name).filter(temp_filter)
+    #sig, read_df = df_read_retry(sql_context, temp_table_name, temp_filter)
+    #assert sig == True
+    #assert read_df != []
+
+    test_row_out = read_df.collect()[0]
+
+    assert test_row_out['field1'] == 'field1_val'
+    assert test_row_out['field2'] == 'field2_val'
+    assert test_row_out['datetime'] == datetime.datetime.fromtimestamp(1293840000)
+    assert test_row_out['data'] == 0
+
+
+    temp_filter = """datetime >= %(start_date)s
+                    AND datetime <=  %(end_date)s
+                    AND field1 = '%(field1)s'
+                    AND field2 = '%(field2)s'
+                """ % ({'start_date': 1293840000000-1, 'end_date': 1293840000000+1, 'field1': 'field1_val', 'field2': 'field2_val'})
+
+    read_df = sql_context.read.format("org.apache.spark.sql.riak").option("spark.riakts.bindings.timestamp", "useLong").load(temp_table_name).filter(temp_filter)
+    #sig, read_df = df_read_retry(sql_context, temp_table_name, temp_filter, useLong = True)
+    #assert sig == True
+    #assert read_df != []
+
+    test_row_out = read_df.collect()[0]
+    
+    assert test_row_out['field1'] == 'field1_val'
+    assert test_row_out['field2'] == 'field2_val'
+    assert test_row_out['datetime'] == 1293840000000
+    assert test_row_out['data'] == 0
```

Comment:
This and other tests are long. The general rule is AAA:

1. Assemble - do whatever to meet the preconditions of the functionality under test
2. Act - do the action, likely changing state (considering a returned value as changing the client's state)
3. Assert - test that the post-condition(s) are met
3.a. To reduce the overall test suite's Assemble code / execution, use setup/teardown and/or functions (as in DRY).